share/wake/lib/rust_wake/cargo.wake

```diff
@@ -25,21 +29,193 @@ export target findCargoPath Unit =
+    CargoToolchainBuildConfigRelease
 
-    require Pass outputs =
-        makeExecPlan (cargoPath, "build", "--release", "--locked", Nil) sources
-        | setPlanLabel "cargo: {pathToCrate}"
-        | setPlanDirectory pathToCrate
-        | setPlanFnOutputs (\_ map ("{pathToCrate}/target/release/{_}") outputs)
-        | prependPlanPath (dirname cargoPath)
+# An on-disk cargo crate
+#
+# Used to specify the root/main crate for the build and any other on-disk crate deps
+export tuple CargoCrate =
+    # Root directory of the crate 
+    export Root: Path
+    # Name of the binary to be build
+    export BinaryName: String
+
+# The cargo build toolchain
+#
+# Describes a specific rust binary to be built using cargo
+export tuple CargoToolchain =
+    # Path to cargo binary
+    export Cargo: String
+    # Build target to use
+    export BuildConfig: CargoToolchainBuildConfig
+    # Root level crate to build
+    export RootCrate: CargoCrate
+    # List of dependancies to crates on disk
+    export LocalCrates: List CargoCrate
+    # Force cargo to use the Cargo.lock file for reproducible builds
+    export Locked: Boolean
+    # Extra cargo flags defined by caller
+    export Flags: List String
+
+# Makes a simple default cargo toolchain for locked release builds.
+#
+# Relies on the system install of cargo so it is not reproducable across machines. For a more reproducable
+# option, use makeCargoToolchain
+export def makeReleaseCargoToolchain (root: Path) (binaryName: String): Result CargoToolchain Error =
+    require Pass cargo = findCargoPath Unit
+
+    makeCargoToolchain cargo root binaryName CargoToolchainBuildConfigRelease True
+    | Pass
+
+# Makes a cargo toolchain without assuming any defaults. Local crates and extra flags must be manually set later
+export def makeCargoToolchain (cargo: String) (root: Path) (binaryName: String) (config: CargoToolchainBuildConfig) (locked: Boolean): CargoToolchain =
+    CargoToolchain cargo config (CargoCrate root binaryName) Nil locked Nil
+
+# Adds a crate to the toolchains list of local crates. Chain calls together to build up a binary
+# with local crate deps.
+#
+# Ex:
+# ```
+# require Pass foo = mkdir "foo"
+# require Pass crate1 = mkdir "foo"
+# require Pass crate2 = mkdir "foo"
+# require Pass toolchain = makeReleaseCargoToolchain foo "bar" 
+# toolchain
+# | addCargoCrate crate1 "one"
+# | addCargoCrate crate2 "two"
+# ```
+export def addCargoCrate (root: Path) (binaryName: String): CargoToolchain => CargoToolchain =
+    editCargoToolchainLocalCrates ((CargoCrate root binaryName), _)
+
+# Adds a flag to the list of flags. Will be forwared to cargo in order after all other flags
+#
+# Ex:
+# ```
+# require Pass foo = mkdir "foo"
+# require Pass toolchain = makeReleaseCargoToolchain foo "bar" 
+# toolchain
+# | addCargoFlag "--foo"
+# | addCargoFlag "bar"
+# | addCargoFlag "--bat"
+# ```
+export def addCargoFlag (flag: String): CargoToolchain => CargoToolchain =
```

Comment:
The `add` verbiage matches what we do for the http and argparse libraries so personally I'd prefer to keep it. Right now we have the mapping 

`setFoo` -> overwrite foo with new value
`addFoo` -> add a new item to Foo which is probably a list

Over time I want us to get in the habit of writing these small helper functions in our "user consumable libraries" as it only takes us a little bit of work while massively improving the DX of the library. I also think we can smooth over the the annoyance of `getTypeNameVarName` if we carefully choose/implement certain short names. This is one of those cases where the compiler can never insert short names because it would generate conflicting names but we can since we know certain packages wont be used together or certain names (like cargo) are unlikely to come up in other types.